# Studio: Working with Databases in Python

For today's studio, we will be using the [TV Shows dataset](https://www.kaggle.com/ruchi798/tv-shows-on-netflix-prime-video-hulu-and-disney) from Kaggle. We have already downloaded the CSV for you.

You will use the watchlist you created to answer these questions:

1. **Which streaming services contain the shows you want to watch next?**
2. **Which streaming service is the best value based on the shows you want to watch?**

As you complete the different tasks in the studio, you may choose between using Pandas or SQL. 

**Remember**: we learned in our prep work that one is oftentimes more efficient at certain tasks than the other, so choose wisely!

## My Watchlist

If you would like, please use this space to make note of your watchlist by editing the text cell. You will need 10 shows overall.

1. Jet Lag
2. The Grand Tour
3. Economics Explained
4. Johnny Harris
5. RealLifeLore
6. Half as Interesting
7. Paddy Doyle
8. Kara and Nate
9. Bald and Bankrupt
10. Ted Lasso

## Database Setup

Import the necessary libraries and create a dataframe from the provided CSV. 

Print the info out for the dataframe. 

After that, you may drop the column called `Unnamed: 0` and rename any columns with spaces or unusual characters in the names such as `"Disney+"`. 

Print out the info for the dataframe again to ensure your changes were made.

In [4]:
# Code here

import sqlite3
import pandas as pd

df = pd.read_csv(r'C:\Users\bccut\OneDrive\Desktop\LaunchCode\data-analysis-projects\databases-python-pandas\studio\tv_shows.csv')

df.drop('Unnamed: 0', axis=1, inplace=True)

print(df.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5368 entries, 0 to 5367
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   ID               5368 non-null   int64 
 1   Title            5368 non-null   object
 2   Year             5368 non-null   int64 
 3   Age              3241 non-null   object
 4   IMDb             4406 non-null   object
 5   Rotten Tomatoes  5368 non-null   object
 6   Netflix          5368 non-null   int64 
 7   Hulu             5368 non-null   int64 
 8   Prime Video      5368 non-null   int64 
 9   Disney+          5368 non-null   int64 
 10  Type             5368 non-null   int64 
dtypes: int64(7), object(4)
memory usage: 461.4+ KB
None


With your dataframe at the ready, create a new database called `tv.db`. 

Add a new table to your database called `shows` using the data in the dataframe. 

In [ ]:
# Code here
tv_db = sqlite3.connect('tv.db')
df.to_sql('shows', tv_db, if_exists='replace')
tv_db.execute(
    '''
    create table shows as
    select * from shows
    '''
)


With your new table and database set up, print out the top 20 records in the `shows` table.

In [12]:
# Code Here
# Create a cursor object
cur = tv_db.cursor()

# Execute the SQL query
cur.execute("SELECT * FROM shows LIMIT 20;")

# Fetch the results
results = cur.fetchall()

# Print the results
for row in results:
    print(row)


(0, 1, 'Breaking Bad', 2008, '18+', '9.4/10', '100/100', 1, 0, 0, 0, 1)
(1, 2, 'Stranger Things', 2016, '16+', '8.7/10', '96/100', 1, 0, 0, 0, 1)
(2, 3, 'Attack on Titan', 2013, '18+', '9.0/10', '95/100', 1, 1, 0, 0, 1)
(3, 4, 'Better Call Saul', 2015, '18+', '8.8/10', '94/100', 1, 0, 0, 0, 1)
(4, 5, 'Dark', 2017, '16+', '8.8/10', '93/100', 1, 0, 0, 0, 1)
(5, 6, 'Avatar: The Last Airbender', 2005, '7+', '9.3/10', '93/100', 1, 0, 1, 0, 1)
(6, 7, 'Peaky Blinders', 2013, '18+', '8.8/10', '93/100', 1, 0, 0, 0, 1)
(7, 8, 'The Walking Dead', 2010, '18+', '8.2/10', '93/100', 1, 0, 0, 0, 1)
(8, 9, 'Black Mirror', 2011, '18+', '8.8/10', '92/100', 1, 0, 0, 0, 1)
(9, 10, "The Queen's Gambit", 2020, '18+', '8.6/10', '92/100', 1, 0, 0, 0, 1)
(10, 11, 'Mindhunter', 2017, '18+', '8.6/10', '90/100', 1, 0, 0, 0, 1)
(11, 12, 'Community', 2009, '7+', '8.5/10', '90/100', 1, 1, 1, 0, 1)
(12, 13, 'Narcos', 2015, '18+', '8.8/10', '90/100', 1, 0, 0, 0, 1)
(13, 14, 'Shameless', 2011, '18+', '8.5/10', '90/100',

Now, create a new table called `watchlist` that has three fields:
1. id -> data type of `INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT`
2. title -> data type of `TEXT`
3. importance_rank -> data type of `INTEGER`

For the `importance_rank` field, rank each of your watchlist shows based on how much you want to see them, `10` being the most important and `1` being the least important.

Then, insert each of the items from your watchlist into the new `watchlist` table, using the `executemany` method from our exercises.

Finally, select all the records from the `watchlist` table and print them out to the console.

In [17]:
# Code here

# cur.execute("CREATE TABLE watchlist (id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT, title TEXT, importance_rank INTEGER)")

# cur.execute("""
#     BEGIN;
#     INSERT INTO watchlist ('Jet Lag', 1);
#     INSERT INTO watchlist ('Grand Tour', 2);
#     INSERT INTO watchlist ('Economics Explained', 3);
#     INSERT INTO watchlist ('Johnny Harris', 4);
#     INSERT INTO watchlist ('RealLifeLore', 5);
#     INSERT INTO watchlist ('Half as Interesting', 6);
#     INSERT INTO watchlist ('Paddy Doyle', 7);
#     INSERT INTO watchlist ('Kara and Nate', 8);
#     INSERT INTO watchlist ('Bald and Bankrupt', 9);
#     INSERT INTO watchlist ('Indigo Traveller', 10);
#     COMMIT;
# """)


# cur.executemany("INSERT INTO watchlist (title, importance_rank) VALUES (?, ?)", watchlist_data)

# Commit the changes
tv_db.commit()

# Select and print all records from the 'watchlist' table
cur.execute("SELECT * FROM watchlist")
results = cur.fetchall()
for row in results:
    print(row)

## Working with the Data

Using Pandas or SQL, find the answer to these 2 questions:
1. How many of the total shows (full csv list) are on each streaming service? 5368
2. What percentage of these total shows is available on each streaming service?

**Hint**:

Use the pandas `query` method to filter the data, and then the Python `len` method to find it's length. [Relevant Link](https://www.geeksforgeeks.org/ways-to-filter-pandas-dataframe-by-column-values/)

In [53]:
# Code here

total_shows = len(df)
print(total_shows)

shows_per_netflix = df.groupby('Netflix').size()
percentage_netflix = (shows_per_netflix / total_shows) * 100

shows_per_hulu = df.groupby('Hulu').size()
percentage_hulu = (shows_per_hulu / total_shows) * 100

shows_per_pv = df.groupby('Prime Video').size()
percentage_pv = (shows_per_pv / total_shows) * 100

shows_per_disney = df.groupby('Disney+').size()
percentage_disney = (shows_per_disney / total_shows) * 100

print("Number of Shows per Streaming Service:")
print(shows_per_netflix)
print(shows_per_hulu)
print(shows_per_pv)
print(shows_per_disney)

print("\nPercentage of Shows per Streaming Service:")
print(percentage_netflix)
print(percentage_hulu)
print(percentage_pv)
print(percentage_disney)


5368
Number of Shows per Streaming Service:
Netflix
0    3397
1    1971
dtype: int64
Hulu
0    3747
1    1621
dtype: int64
Prime Video
0    3537
1    1831
dtype: int64
Disney+
0    5017
1     351
dtype: int64

Percentage of Shows per Streaming Service:
Netflix
0    63.282414
1    36.717586
dtype: float64
Hulu
0    69.802534
1    30.197466
dtype: float64
Prime Video
0    65.890462
1    34.109538
dtype: float64
Disney+
0    93.461252
1     6.538748
dtype: float64



Now join your `watchlist` data to the `shows` data using pandas or SQL. Verify that you joined the data correctly.

Using this related dataset, come up with analytic code that answers these questions:
1. The number of watchlist shows each streaming service has
2. The percentage of your overall watchlist each streaming service has


In [ ]:
# Code here
conn = sqlite3.connect('tv.db')

conn = sqlite3.connect('watchlist.db')

# Read the tables into DataFrames
table1_df = pd.read_sql_query("SELECT * FROM shows", conn)
table2_df = pd.read_sql_query("SELECT * FROM watchlist", conn)

# Join the DataFrames on a common column (replace 'common_column' with the actual column name)
merged_df = pd.merge(table1_df, table2_df, on='Title')

# Print the merged DataFrame
print(merged_df)

## Results

Now that you have done your analysis, make note of the answers to the following questions by editing the text cell:

1. Was every show on your watchlist in the Kaggle dataset? Do you have any ideas as to why a show might not have been present?

- No, not every show was in the Kaggle Dataset.  This is because they don't have an agreement with one of the streaming services. 

2. Did you include a show or shows in your watchlist that is exclusive to one of the platforms? How might that have impacted your analysis?

- Yes, The Grand Tour.  The only one that would show up if inner joined on Title.

3. Which streaming service(s) offered the most shows on your watchlist? Which streaming service(s) offered the least?

- Prime offered the most with 1 show.  Netflix, Hulu, and Disney offered the least.

4. Based on the shows you want to watch and the results of your analysis, is there a streaming service you think would be a good fit for you?

- None would be a good fit.  Since there is not much tv I would like to watch no use in having a streaming service.  

# Bonus Mission

We didn't end up using that `importance_rank` field, did we?

Well, that was intentional! 

Your bonus mission is to come up with analysis that uses that field to determine, based on watchlist show importance_rank and number of watchlist shows available on a service, which platform you should subscribe to.

In [ ]:
# Code Here